src/mesh.cpp

```diff
@@ -0,0 +1,981 @@
+  // We are looking for the first valid mesh bin.  Check to see if the
+  // particle starts inside the mesh.
+  if (!start_in_mesh) {
+    xt::xtensor<double, 1> d = xt::zeros<double>({n});
+
+    // The particle does not start in the mesh.  Note that we nudged the
+    // start and end coordinates by a TINY_BIT each so we will have
+    // difficulty resolving tracks that are less than 2*TINY_BIT in length.
+    // If the track is that short, it is also insignificant so we can
+    // safely ignore it in the tallies.
+    if (total_distance < 2*TINY_BIT) return;
+
+    // The particle does not start in the mesh so keep iterating the ijk0
+    // indices to cross the nearest mesh surface until we've found a valid
+    // bin.  MAX_SEARCH_ITER prevents an infinite loop.
+    int search_iter = 0;
+    int j;
+    while (xt::any(ijk0 < 1) || xt::any(ijk0 > shape_)) {
+      if (search_iter == MAX_SEARCH_ITER) {
+        warning("Failed to find a mesh intersection on a tally mesh filter.");
+        return;
+      }
+
+      for (j = 0; j < n; ++j) {
+        if (std::fabs(u[j]) < FP_PRECISION) {
+          d(j) = INFTY;
+        } else if (u[j] > 0.0) {
+          double xyz_cross = lower_left_[j] + ijk0(j) * width_[j];
+          d(j) = (xyz_cross - r0[j]) / u[j];
+        } else {
+          double xyz_cross = lower_left_[j] + (ijk0(j) - 1) * width_[j];
+          d(j) = (xyz_cross - r0[j]) / u[j];
+        }
+      }
+
+      j = xt::argmin(d)(0);
+      if (u[j] > 0.0) {
+        ++ijk0(j);
+      } else {
+        --ijk0(j);
+      }
+
+      ++search_iter;
+    }
+
+    // Advance position
+    r0 += d(j) * u;
+  }
+
+  while (true) {
+    // ========================================================================
+    // Compute the length of the track segment in the appropiate mesh cell and
+    // return.
+
+    double distance;
+    int j;
+    if (ijk0 == ijk1) {
+      // The track ends in this cell.  Use the particle end location rather
+      // than the mesh surface.
+      distance = (r1 - r0).norm();
+    } else {
+      // The track exits this cell.  Determine the distance to the closest mesh
+      // surface.
+      xt::xtensor<double, 1> d = xt::zeros<double>({n});
+      for (int j = 0; j < n; ++j) {
+        if (std::fabs(u[j]) < FP_PRECISION) {
+          d(j) = INFTY;
+        } else if (u[j] > 0) {
+          double xyz_cross = lower_left_[j] + ijk0(j) * width_[j];
+          d(j) = (xyz_cross - r0[j]) / u[j];
+        } else {
+          double xyz_cross = lower_left_[j] + (ijk0(j) - 1) * width_[j];
+          d(j) = (xyz_cross - r0[j]) / u[j];
+        }
+      }
+      j = xt::argmin(d)(0);
+      distance = d(j);
+    }
+
+    // Assign the next tally bin and the score.
+    int bin = get_bin_from_indices(ijk0.data());
+    bins.push_back(bin);
+    lengths.push_back(distance / total_distance);
+
+    // Find the next mesh cell that the particle enters.
+
+    // If the particle track ends in that bin, { we are done.
```

Comment:
`that` -> `this` and remove the `{`

Reply:
Haha, got a little overzealous with my `then` -> `{` replacements